File: web/lib/dividends/merkle.ts
// Merkle tree for dividend epochs, byte-compatible with programs/causa_vault/src/merkle.rs (Jito distributor scheme):
//   leaf = keccak256(0x00 ‖ epochId LE u64 ‖ account (32) ‖ n (u8) ‖ amounts[i] LE u64 …)
//   node = keccak256(0x01 ‖ min(a, b) ‖ max(a, b))
// Leaves are ordered by account so the same input always yields the same root; an odd node is promoted unchanged.
import { keccak_256 } from "@noble/hashes/sha3.js";
import { PublicKey } from "@solana/web3.js";

export interface TreeLeaf {
  account: string;
  amounts: bigint[];
}

export interface TreeDump {
  epochId: string;
  root: string;
  leaves: { account: string; amounts: string[] }[];
}

export interface BuiltTree {
  root: `0x${string}`;
  /** account -> proof (hex nodes) */
  proofs: Map<string, `0x${string}`[]>;
  /** account -> leaf hash */
  leafHashes: Map<string, `0x${string}`>;
  dump: TreeDump;
}

// DataView rather than Buffer.writeBigUInt64LE so the same code runs under the browser Buffer polyfill.
const u64le = (v: bigint) => {
  const b = Buffer.alloc(8);
  new DataView(b.buffer, b.byteOffset, 8).setBigUint64(0, v, true);
  return b;
};

export const hex = (b: Uint8Array): `0x${string}` => `0x${Buffer.from(b).toString("hex")}`;
export const unhex = (h: string): Buffer => Buffer.from(h.replace(/^0x/, ""), "hex");

export function leafHash(epochId: bigint, account: string, amounts: bigint[]): Uint8Array {
  return keccak_256(
    Buffer.concat([
      Buffer.from([0]),
      u64le(epochId),
      new PublicKey(account).toBuffer(),
      Buffer.from([amounts.length]),
      ...amounts.map(u64le),
    ]),
  );
}

function nodeHash(a: Uint8Array, b: Uint8Array): Uint8Array {
  const [lo, hi] = Buffer.compare(Buffer.from(a), Buffer.from(b)) <= 0 ? [a, b] : [b, a];
  return keccak_256(Buffer.concat([Buffer.from([1]), Buffer.from(lo), Buffer.from(hi)]));
}

export function buildTree(epochId: bigint, input: TreeLeaf[]): BuiltTree {
  if (input.length === 0) throw new Error("merkle: no leaves");
  const leaves = [...input].sort((a, b) => a.account.localeCompare(b.account));
  const levels: Uint8Array[][] = [leaves.map((l) => leafHash(epochId, l.account, l.amounts))];
  while (levels[levels.length - 1].length > 1) {
    const prev = levels[levels.length - 1];
    const next: Uint8Array[] = [];
    for (let i = 0; i < prev.length; i += 2) next.push(i + 1 < prev.length ? nodeHash(prev[i], prev[i + 1]) : prev[i]);
    levels.push(next);
  }
  const root = levels[levels.length - 1][0];
  const proofs = new Map<string, `0x${string}`[]>();
  const leafHashes = new Map<string, `0x${string}`>();
  leaves.forEach((leaf, index) => {
    const proof: `0x${string}`[] = [];
    let i = index;
    for (let lvl = 0; lvl < levels.length - 1; lvl++) {
      const sibling = i ^ 1;
      if (sibling < levels[lvl].length) proof.push(hex(levels[lvl][sibling]));
      i = Math.floor(i / 2);
    }
    proofs.set(leaf.account, proof);
    leafHashes.set(leaf.account, hex(levels[0][index]));
  });
  return {
    root: hex(root),
    proofs,
    leafHashes,
    dump: { epochId: epochId.toString(), root: hex(root), leaves: leaves.map((l) => ({ account: l.account, amounts: l.amounts.map(String) })) },
  };
}

export function verifyLeaf(root: string, epochId: bigint, leaf: TreeLeaf, proof: string[]): boolean {
  let node = leafHash(epochId, leaf.account, leaf.amounts);
  for (const p of proof) node = nodeHash(node, unhex(p));
  return Buffer.from(node).equals(unhex(root));
}

File: solana/programs/causa_vault/src/merkle.rs
//! Merkle leaves and proofs for dividend epochs (the Jito distributor scheme, keccak-based):
//!   leaf = keccak256(0x00 ‖ epoch_id LE ‖ account ‖ n ‖ amounts[i] LE …)
//!   node = keccak256(0x01 ‖ min(a, b) ‖ max(a, b))
//! The TypeScript builder in web/lib/dividends/merkle.ts must produce identical bytes.
use solana_keccak_hasher::hashv;

pub const LEAF_PREFIX: &[u8] = &[0u8];
pub const NODE_PREFIX: &[u8] = &[1u8];

pub fn leaf_hash(epoch_id: u64, account: &[u8; 32], amounts: &[u64]) -> [u8; 32] {
    let epoch = epoch_id.to_le_bytes();
    let n = [amounts.len() as u8];
    let amount_bytes: Vec<u8> = amounts.iter().flat_map(|a| a.to_le_bytes()).collect();
    hashv(&[LEAF_PREFIX, &epoch, account, &n, &amount_bytes]).to_bytes()
}

pub fn verify(proof: &[[u8; 32]], root: &[u8; 32], leaf: [u8; 32]) -> bool {
    let mut node = leaf;
    for sibling in proof {
        node = if node <= *sibling {
            hashv(&[NODE_PREFIX, &node, sibling]).to_bytes()
        } else {
            hashv(&[NODE_PREFIX, sibling, &node]).to_bytes()
        };
    }
    node == *root
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn single_leaf_tree_root_is_leaf() {
        let leaf = leaf_hash(1, &[7u8; 32], &[10, 0, 5]);
        assert!(verify(&[], &leaf, leaf));
    }

    #[test]
    fn two_leaf_tree() {
        let a = leaf_hash(1, &[1u8; 32], &[1]);
        let b = leaf_hash(1, &[2u8; 32], &[2]);
        let (lo, hi) = if a <= b { (a, b) } else { (b, a) };
        let root = hashv(&[NODE_PREFIX, &lo, &hi]).to_bytes();
        assert!(verify(&[b], &root, a));
        assert!(verify(&[a], &root, b));
        assert!(!verify(&[a], &root, a));
    }
}
